### ISSUE TITLE
Data can persist?

### ISSUE THREAD

**Issue by q96456** [facebookincubator/beringei#26]
Data only in memory?

**Comment by kiranisaac** [comment]
Data is persisted to disk to allow the service to restart with minimal data loss. The current version does not support a long term external data store. This should be easy to extend by reading from Beringei and writing to a permanent storage.

**Comment by at15** [comment]
I think the Gorilla paper said data is written to HBase for long term storage, will Beringei consider adding the capability of writing to other storage or it should be done externally?

**Comment by kiranisaac** [comment]
Within Facebook, we read the data out of Beringei and write it to HBase. This feature is part of our long term roadmap. For the short term, we are more interested integrating with stats collection tools. 

If the community feels like persistent long term storage is very important, we are will adjust our plan accordingly.

cc: @jteller

**Comment by q96456** [comment]
Can I see it as a recent data of snapshot?

**Comment by kiranisaac** [comment]
@q96456, I am sorry, I don't understand the question. Could you rephrase it?

**Comment by mlgood** [comment]
@q96456 is the question whether data in beringei is effectively a snapshot of recent data?

At Facebook, data in Beringei functions as a write-back cache of the most recent 24 hours worth of data.  As @kiranisaac mentioned above, within Facebook we read the data out of Beringei and write it back to HBase.  What we've open sourced here can, at any point in time, be seen as a snapshot of the previous 24 hours' data, though it will be continuously changing as data ages out and new data arrives.

**Comment by hshopeful** [comment]
@mlgood in the paper, it claims that "Gorilla functions as a write-through cache of the most recent data entering the monitoring system", but in your above answer, you says that "Beringei functions as a write-back cache of the most recent 24 hours worth of data".
write-through or write-back, which one is accurate?

**Comment by jteller** [comment]
@hshopeful - The answer to your question is "yes"

When we wrote the paper, it was implemented as a write-through cache (because the process that wrote to Gorilla also wrote to HBase long term storage). Now, it's used as a write-back cache (all data is streamed into Gorilla/Beringei and a different service reads from Gorilla/Beringei and writes to HBase). We made the switch because it allowed for more efficient batching to the HBase store, and our history operating the system had shown that Gorilla/Beringei was reliable enough to act as a write-back cache. Especially because we maintain several peer Gorillas.

**Comment by hshopeful** [comment]
@jteller thanks for your answer. I have another questions.

1、In the interior of Gorilla, it writes data in-momory, it also persists data in disk almost at the same time, can I understand that the write policy in the interior of Gorilla is write-through?

2、In the source code, I can not find the "ShardManager" which mentioned in the paper, and I find something relative to it in "BeringeiConfigurationAdapter.cpp", unluckily, I can not understand it thoroughly. There is a configuration file in each node(assigns which shards are stored in it), how does the configuration file on each nodes keep synchronization and collaborative. Can you explain the details about "adding nodes to the cluster" and "remove nodes from the cluster", and point out where it is in the code

3、In Gorrila, it uses GlusterFS to do data persistence which mentioned in the paper, but I can not find the code about "GlusterFS", can you point out where it is in the code

Thanks

**Comment by jteller** [comment]
@hshopeful - good questions! Based on them though, it sounds like the original question was answered.

1. Beringei/Gorilla disk persistence is like a "write behind" log. It buffers a small amount of data in memory before flushing to disk. On catastrophic failure, data buffered but not flushed to disk will be lost - but on graceful shutdown no data will be lost. In practice, the performance boost from buffering additional writes was worth the data loss on crashes. BucketLogWriter.cpp is the file that writes this log to disk. logData() is called on every write, and simply places the data into a queue. Another thread calls writeOneLogEntry() to pull out of the queue and write to disk

2. ShardManager is a Facebook internal system - it is not open source, so we couldn't depend on it in Beringei. Keeping the configuration file in sync on all the systems is an exercise left to the reader at this point: we do not include any code or system to do that. However, using something like Zookeeper to distribute and synchronize the shard map can work. The code in Beringei doesn't explicitly handle adding new nodes to a cluster or removing nodes.

3. GlusterFS is a POSIX filesystem that can run on a cluster. Interfacing with it is similar to NFS - so we simply mount the gluster cluster on each node, and point the argument "data_directory" to the mounted file system. Any place you see Beringei writing or reading from a file, we expect those files to be stored on some sort of shared file system if you want to allow persistence across single-node failures.

**Comment by jteller** [comment]
I'm going to close this issue as resolved as I believe we have answered the original question
